### PR TITLE
Web3 integration addressbook

### PR DESCRIPTION
**Pull request by chadoh**
# AddressBook web3 integration

## What was done

- Merged changes strictly related to AddressBook mainnet audit, its contracts and structure, tests and packages
- Web3 integration into the AddressBook store adapting it to the contract changes
- Minor fix after rebasing AddressBook frontend to allocate the new `entry` object schema
- The `updateEntry` is implemented as a function in `App.js` and tracked in the store, we only need to create the `ContextMenuItem` for it and the SidePanel with the form to manage the edit action (I skipped this part because it is out of the scope)  
 
It should be ready to merge

### CODE REVIEW

File: apps/address-book/app/store/entry.js [hunk 1]
```diff
@@ -29,20 +30,43 @@ export const onEntryRemoved = async ({ entries }, { addr }) => {
...
-    app.call('getEntry', addr).subscribe(entry => {
-      // don't resolve when entry not found
-      entry &&
-        resolve({
-          entryAddress: entry[0],
-          name: entry[1],
-          entryType: entry[2],
+  return new Promise((resolve, reject) => {
+    app.call('getEntry', addr).subscribe(async cid => {
+      if (!cid) resolve() // entry probably removed in a future block
```
Comment (topocount) [inline apps/address-book/app/store/entry.js]: This isn't behaving as expected when handling events for entries that have already been removed. IPFS reolution errors are throwing from [line 59](https://github.com/AutarkLabs/open-enterprise/pull/1189/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.sol&file-filters%5B%5D=dotfile&hide-deleted-files=true#diff-446ecc1afaa88101fd8bca756c71b84aR59)

File: apps/address-book/app/store/entry.js [hunk 2]
```diff
@@ -51,21 +51,15 @@ export const onEntryUpdated = async ({ entries = [] }, { addr }) => {
...
     app.call('getEntry', addr).subscribe(async cid => {
-      if (!cid) resolve() // entry probably removed in a future block
-      try {
+      if (!cid) {
+        resolve() // entry probably removed in a future block
+      } else {
         const entryData = await ipfsGet(cid)
         // It is also needed track the cid because the remove function needs it
         resolve({ ...entryData, cid })
-      } catch (error) {
```
Comment (chadoh) [inline apps/address-book/app/store/entry.js]: Are we confident that the `await ipfsGet` will never throw an error, @ottodevs?

File: apps/address-book/app/store/entry.js [hunk 3]
```diff
@@ -51,21 +51,15 @@ export const onEntryUpdated = async ({ entries = [] }, { addr }) => {
 /*    AddressBook helper functions    */
 /// /////////////////////////////////////
 
-const loadEntryData = (addr) => {
-  return new Promise((resolve, reject) => {
+const loadEntryData = addr => {
+  return new Promise(resolve => {
     app.call('getEntry', addr).subscribe(async cid => {
-      if (!cid) resolve() // entry probably removed in a future block
```
Comment (chadoh) [inline apps/address-book/app/store/entry.js]: Interesting that JS will continue to evaluate the function body after `resolve` is called. Good catch.

File: apps/address-book/app/store/entry.js (hunk 2 above)
Comment (chadoh) [reply apps/address-book/app/store/entry.js]: From @ottodevs in Keybase chat:

> `ipfsGet` already has its own error management, why add another one over it?

💯

File: apps/allocations/package.json [hunk 4]
```diff
@@ -80,7 +80,7 @@
   "keywords": [],
   "license": "GPL-3.0",
   "lint-staged": {
-    "*.js": [
+    "app/**/*.js": [
```
Comment (chadoh) [inline apps/allocations/package.json]: Please remove all changes to `apps/allocations`

File: apps/allocations/package-lock.json [hunk 5]
```diff
@@ -0,0 +1,3317 @@
+{
```
Comment (chadoh) [inline apps/allocations/package-lock.json]: Please get rid of this file! We are not committing `package-lock` files.

File: apps/discussions/package.json [hunk 6]
```diff
@@ -33,7 +33,8 @@
     "eslint-plugin-standard": "^4.0.0",
     "prettier": "^1.17.1",
     "react-hot-loader": "^4.8.8",
-    "solidity-coverage": "0.5.11"
+    "solidity-coverage": "0.5.11",
+    "solium": "1.1.8"
```
Comment (chadoh) [inline apps/discussions/package.json]: How did these `discussions` changes sneak into this PR?

File: apps/dot-voting/.eslintrc.js [hunk 7]
```diff
@@ -54,6 +54,31 @@ module.exports = {
...
+          "off",
+          {
+            unusedExports: true,
+            missingExports: true,
+            ignoreExports: [],
+          }
+        ],
+      }
+    }
+  ],
```
Comment (chadoh) [inline apps/dot-voting/.eslintrc.js]: How did `dot-voting` changes sneak into this PR?

File: apps/dot-voting/package.json [hunk 8]
```diff
@@ -81,7 +81,7 @@
   "keywords": [],
   "license": "GPL-3.0",
   "lint-staged": {
-    "*.js": [
+    "app/**/*.js": [
```
Comment (chadoh) [inline apps/dot-voting/package.json]: Please remove.

File: apps/planning-suite-kit/package.json [hunk 9]
```diff
@@ -12,13 +12,13 @@
...
-    "@aragon/apps-voting": "^2.0.0",
-    "@aragon/apps-vault": "^4.1.0",
-    "@aragon/os": "^4.2.0",
+    "@aragon/apps-finance": "3.0.0",
+    "@aragon/apps-shared-minime": "1.0.1",
+    "@aragon/apps-survey": "1.0.0-rc.2",
+    "@aragon/apps-token-manager": "2.0.0",
+    "@aragon/apps-voting": "2.0.0",
+    "@aragon/apps-vault": "4.1.0",
+    "@aragon/os": "4.2.0",
```
Comment (chadoh) [inline apps/planning-suite-kit/package.json]: This is all good stuff but probably unrelated to getting Address Book working. If that's so, please move this to a new PR.

File: apps/projects/.eslintrc.js [hunk 10]
```diff
@@ -51,6 +51,31 @@ module.exports = {
...
+          "off",
+          {
+            unusedExports: true,
+            missingExports: true,
+            ignoreExports: [],
+          }
+        ],
+      }
+    }
+  ],
```
Comment (chadoh) [inline apps/projects/.eslintrc.js]: Please remove.

File: apps/projects/package-lock.json [hunk 11]
```diff
@@ -0,0 +1,642 @@
+{
```
Comment (chadoh) [inline apps/projects/package-lock.json]: Please remove this file.

File: apps/projects/package.json [hunk 12]
```diff
@@ -87,7 +86,7 @@
   "keywords": [],
   "license": "GPL-3.0",
   "lint-staged": {
-    "*.js": [
+    "app/**/*.js": [
```
Comment (chadoh) [inline apps/projects/package.json]: Please remove all `apps/projects/package.json` changes from this PR

File: apps/rewards/truffle-config.js [hunk 13]
```diff
@@ -1 +1 @@
-module.exports = require('@tps/test-helpers/truffle-config')
\ No newline at end of file
+module.exports = require("@aragon/os/truffle-config")
```
Comment (chadoh) [inline apps/rewards/truffle-config.js]: Please remove all `apps/rewards` changes from this PR.

File: kits/planning-suite/contracts/PlanningSuite.sol [hunk 14]
```diff
@@ -463,8 +463,8 @@ contract PlanningSuite is BetaKitBase {
     {
         ACL acl = ACL(dao.acl());
         bytes32 vaultTransferRole = vault.TRANSFER_ROLE();
-        // Vault permissions
 
+        // Vault permissions
```
Comment (chadoh) [inline kits/planning-suite/contracts/PlanningSuite.sol]: Let's get rid of this noise from this PR.

File: kits/planning-suite/package.json [hunk 15]
```diff
@@ -48,24 +48,24 @@
...
     "@tps/apps-allocations": "^0.0.1",
     "@tps/apps-projects": "^0.0.1",
     "@tps/apps-dot-voting": "^0.0.1",
     "@tps/apps-rewards": "^0.0.1"
   },
   "devDependencies": {
-    "@aragon/id": "^2.0.3",
-    "@aragon/test-helpers": "^1.0.0",
+    "@aragon/id": "2.0.3",
+    "@aragon/test-helpers": "^1.0.1",
```
Comment (chadoh) [inline kits/planning-suite/package.json]: Do these changes make sense as part of this PR?

File: package.json [hunk 16]
```diff
@@ -25,11 +25,11 @@
...
     "clean:aragon": "rm -rf ~/.aragon && rm -rf ~/.ipfs",
-    "clean:build": "git clean -ffdx --exclude=node_modules/",
+    "clean:build": "npm run clean -- --exclude=node_modules/",
     "clean:commit": "rm -f package-lock.json && npm run bootstrap",
-    "clean": "git clean -ffdx",
-    "coverage": "lerna run coverage --no-bail",
-    "coveralls": "cat apps/*/coverage/lcov.info | coveralls",
+    "clean": "git clean -fXd",
+    "coverage": "./node_modules/.bin/lerna run coverage --no-bail",
+    "coveralls": "cat apps/*/coverage/lcov.info | ./node_modules/.bin/coveralls",
```
Comment (chadoh) [inline package.json]: Good stuff. Should we put it all in its own PR?

File: shared/ui/package-lock.json [hunk 17]
```diff
@@ -0,0 +1,590 @@
+{
```
Comment (chadoh) [inline shared/ui/package-lock.json]: Please remove

File: apps/allocations/.eslintrc.js [hunk 18]
```diff
@@ -54,6 +54,31 @@ module.exports = {
...
+          "off",
+          {
+            unusedExports: true,
+            missingExports: true,
+            ignoreExports: [],
+          }
+        ],
+      }
+    }
+  ],
```
Comment (chadoh) [inline apps/allocations/.eslintrc.js]: Please remove

File: apps/address-book/package.json [hunk 19]
```diff
@@ -76,7 +88,7 @@
   "keywords": [],
   "license": "GPL-3.0",
   "lint-staged": {
-    "*.js": [
+    "app/**/*.js": [
```
Comment (chadoh) [inline apps/address-book/package.json]: We want this line as it was

File: apps/address-book/.eslintrc.js [hunk 20]
```diff
@@ -57,6 +57,31 @@ module.exports = {
...
+          "off",
+          {
+            unusedExports: true,
+            missingExports: true,
+            ignoreExports: [],
+          }
+        ],
+      }
+    }
+  ],
```
Comment (chadoh) [inline apps/address-book/.eslintrc.js]: Do we want these changes as part of this PR, or is this something we should add across all apps in a separate PR?

File: apps/address-book/.gitignore [hunk 21]
```diff
@@ -1,4 +1,30 @@
...
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+
+# built assets
+/public/aragon-ui
+/public/script.js
+/public/script.map
+
+package-lock.json
```
Comment (chadoh) [inline apps/address-book/.gitignore]: Same question for this. It seems helpful, but I'd prefer a repo-wide PR that implements this change separate from the significant effort of upgrading the Address Book contracts.

File: apps/address-book/.soliumrc.json [hunk 22]
```diff
@@ -1,37 +1,22 @@
...
-    "deprecated-suicide": "error",
-    "pragma-on-top": "error",
-    "function-order": "error",
-    "no-constant": "error",
-    "value-in-payable": "error",
-    "max-len": "error",
-    "visibility-first": "error",
-    "linebreak-style": "error"
+    "pragma-on-top": ["error"],
+    "security/enforce-explicit-visibility": ["error"]
```
Comment (chadoh) [inline apps/address-book/.soliumrc.json]: Do we want to upgrade this right now, or was this part of the upgrade to `ethlint`, which we have now delayed?

File: apps/address-book/app/components/App/Entities.js [hunk 23]
```diff
@@ -33,31 +33,30 @@ const Entities = ({ entities, onNewEntity, onRemoveEntity }) => {
   } else {
     return (
       <DataView
-        mode="table"
-        fields={[ 'Entity', '' ]}
+        mode="adaptive"
+        fields={[ 'Name', 'Address', 'Type' ]}
```
Comment (chadoh) [inline apps/address-book/app/components/App/Entities.js]: This PR is now out of date. These changes were part of `addressbook-design-check`, which has now been merged to `dev` (#1214). Please rebase this branch with `dev` so that it's more clear what changed as part of the contract upgrade.

File: apps/address-book/app/store/entry.js [hunk 24]
```diff
@@ -29,20 +30,43 @@ export const onEntryRemoved = async ({ entries }, { addr }) => {
...
+      try {
+        const entryData = await ipfsGet(cid)
+        // It is also needed track the cid because the remove function needs it
+        resolve({ ...entryData, cid })
+      } catch (error) {
+        console.error(error) // TODO: incorporate this info into the errorCode & errorMessage below!!
+        reject({
+          caughtError: error,
+          errorCode: 'ipfs_resolution_error',
+          errorMessage: 'could not resolve ipfs blah blah blah; fix this error message in apps/address-book/app/store/entry.js',
```
[the remark this answers is not in the document]
Comment (ottodevs) [reply apps/address-book/app/store/entry.js]: Solved by removing the reject here an prevailing the `ipfsGet` one

File: apps/address-book/app/store/entry.js (hunk 1 above)
Comment (ottodevs) [reply apps/address-book/app/store/entry.js]: Solved by moving the `ipfsGet` part inside an `else` block, so if `cid` is not provided, then the else condition will be skipped, this way we don't have the error anymore

File: apps/allocations/package-lock.json (hunk 5 above)
Comment (ottodevs) [reply apps/allocations/package-lock.json]: Why not just simply do:

```
echo 'package-lock=false' >> .npmrc
echo 'package-lock.json' >> .gitignore
```

I remember to have done this change before, but at some point it was reverted. I think this is a definitive solution, and even easily reversible in the remote case we want to use it again.

File: apps/discussions/package.json (hunk 6 above)
Comment (ottodevs) [reply apps/discussions/package.json]: Probably we merged `dev` branch here at some point to advance on conflict solving, but it should not count as commit anyway... good catch, cleaning it.

File: apps/dot-voting/.eslintrc.js (hunk 7 above)
Comment (ottodevs) [reply apps/dot-voting/.eslintrc.js]: I'll clean anything that is not strictly related to the scope

File: apps/allocations/package-lock.json (hunk 5 above)
Comment (chadoh) [reply apps/allocations/package-lock.json]: Yes, that sounds like a great PR to make off of `dev`! 👍 

But for now, in this PR, let's get rid of all the `package-lock` files directly.

File: apps/address-book/.eslintrc.js (hunk 20 above)
Comment (ottodevs) [reply apps/address-book/.eslintrc.js]: We did the same for other apps independently, but we can definitely isolate these into a new PR and it makes sense for me as well

File: package.json (hunk 16 above)
Comment (ottodevs) [reply package.json]: Why not? I'll do it

File: shared/ui/package-lock.json (hunk 17 above)
Comment (ottodevs) [reply shared/ui/package-lock.json]: Reminder for me: also push PR to remove this pkg from lerna bootstrap

File: apps/address-book/app/components/App/Entities.js (hunk 23 above)
Comment (ottodevs) [reply apps/address-book/app/components/App/Entities.js]: It seems the code you are pointing has something wrong with the DataView and it broke my previous tested working code:

image

I rebased as you suggested, but now I need to put some extra time debugging and fixing the DataView implementation to be able to have a working version of the app

Comment (ottodevs) [reply apps/address-book/app/components/App/Entities.js]: Fixed! It was easier than I thought, we just needed to rename some of the fields to adapt to the new `entry` object format
